Improve date handling for Elasticsearch backend

gsmet opened the pull request:
Bikeshedding for the name of KeepSafeFromAnalyzerBridge very welcome!

gsmet commented:
FYI, I force pushed a few more tests/validation for the calendar case.

DavideD commented:
Wouldn't it be posssible to reuse `StringEncodingDateBridge` and `StringEncodingCalendarBridge`?

gsmet commented:
As the date format is not the same, we would have to make the formatting function overridable.

I'm not sure the code will be better and more maintainable.

DavideD commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/util/impl/ElasticsearchDateHelper.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public final class ElasticsearchDateHelper {
+
+	private static final TimeZone ENCODING_TIME_ZONE = TimeZone.getTimeZone( "UTC" );
```
It probably doesn't make any difference but the @DateBridge annotation javadoc states the we use GMT as timezone and it seems that it is what we do in the Lucene backend as well.

gsmet replied:
It was preexisting in Gunnar's code. Will switch to GMT though.

DavideD replied:
I think it's because Elasticsearch uses UTC by default (if I understand correctly). I'm not sure if this is going to have an impact.

gsmet replied:
Mmmh, no, in fact, it's voluntary.

From the Es doc: "In JSON documents, dates are represented as strings. Elasticsearch uses a set of preconfigured formats to recognize and parse these strings into a long value representing milliseconds-since-the-epoch in UTC.".

In practice, it's the same but we might as well be in line with the doc.

gsmet replied:
In practice, there's no impact as UTC and GMT share the same current time.

Moreover, we only need to be consistent for a given backend and all the manipulations are done with the FieldBridges so we are on the safe side.

DavideD commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/util/impl/ElasticsearchDateHelper.java:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	public static Calendar dateToCalendar(Date date) {
+		Calendar calendar = Calendar.getInstance( ENCODING_TIME_ZONE, Locale.ROOT );
```
in `dateToString` the Locale used is `ENGLISH` here is `ROOT`, is there a reason for it or can we stick with one?

gsmet replied:
Probably inspired by 2 different conversion codes.

Will fix.

gsmet replied:
Mmmh, no, in fact, it's voluntary.

From the Es doc: "In JSON documents, dates are represented as strings. Elasticsearch uses a set of preconfigured formats to recognize and parse these strings into a long value representing milliseconds-since-the-epoch in UTC.".

In practice, it's the same but we might as well be in line with the doc.

DavideD commented on engine/src/main/java/org/hibernate/search/bridge/KeepSafeFromAnalyzerBridge.java:
```diff
@@ -0,0 +1,17 @@
+ *
+ * @author Guillaume Smet
+ */
+public interface KeepSafeFromAnalyzerBridge {
```
if I understand correctly this interface is used in DSL when we shouldn't apply an analyzer (basically is the same as calling `.ignoreAnalyzer()`). Am I right?

Shouldn't this be in the package: `org.hibernate.search.query.dsl[.impl]`

DavideD replied:
As an alternative name I would suggest `IgnoreAnalyzer` or `SkipAnalysis`

gsmet replied:
Yes, you are. Yeah maybe IgnoreAnalyzerBridge is fine and mirrors the ignoreAnalyzer() method.

As for the package, I don't know what is the rule of thumb for an interface used by other public interfaces.

@Sanne ?

DavideD replied:
I didn't realize that the other bridges are in a public package, `org.hibernate.search.query.dsl` is still public so it should do, but let's see what @Sanne  thinks

DavideD commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchCalendarBridge.java:
```diff
@@ -0,0 +1,62 @@
+import org.hibernate.search.backend.elasticsearch.util.impl.ElasticsearchDateHelper;
+import org.hibernate.search.bridge.LuceneOptions;
+
+public class ElasticsearchCalendarBridge extends ElasticsearchDateBridge {
```
Some Javadoc with the format used to store the dates would be nice

DavideD commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchDateBridge.java:
```diff
@@ -0,0 +1,72 @@
+import org.hibernate.search.bridge.TwoWayFieldBridge;
+import org.hibernate.search.bridge.builtin.impl.DateResolutionUtil;
+
+public class ElasticsearchDateBridge implements TwoWayFieldBridge, ParameterizedBridge, KeepSafeFromAnalyzerBridge {
```
Javadoc

DavideD commented:
As a separate issue or commit, shouldn't we apply `KeepSafeFromAnalyzer` to the `StringEncondingDateBridge` for Lucene?

we will need to update the `@DateBridge` and `CalendarBrdige` javadoc as the timezone is now backend dependent and add some explanation in the documentation as well.

I think, that's all for my review, I'll wait for your response on the remarks about `KeepSafeFromAnalyzer`.

gsmet commented:
> As a separate issue or commit, shouldn't we apply KeepSafeFromAnalyzer to the StringEncondingDateBridge for Lucene?

Yes, forgot this one. Done.

> we will need to update the @DateBridge and CalendarBrdige javadoc as the timezone is now backend dependent and add some explanation in the documentation as well.

I just changed GMT to GMT/UTC. As it's in practice the same thing, I think it will be more noise to add something about the difference between backends.

> I think, that's all for my review, I'll wait for your response on the remarks about KeepSafeFromAnalyzer.

OK, fixed all your remarks.

Waiting for @Sanne to decide where we put the IgnoreAnalyzerBridge interface.

DavideD commented:
>> As a separate issue or commit, shouldn't we apply KeepSafeFromAnalyzer to the StringEncondingDateBridge for Lucene?

> Yes, forgot this one. Done.

In this case you should also add `StringEncondingCalendarBridge`  I guess

gsmet commented:
> In this case you should also add StringEncondingCalendarBridge I guess

No, it extends StringEncodingDateBridge so we're good.

gunnarmorling commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchBridgeProvider.java:
```diff
@@ -6,76 +6,103 @@
+	public FieldBridge provideFieldBridge(ExtendedBridgeProviderContext context) {
+		AnnotatedElement annotatedElement = context.getAnnotatedElement();
+		if ( Date.class.isAssignableFrom( context.getReturnType() ) ) {
+			if ( annotatedElement.isAnnotationPresent( org.hibernate.search.annotations.DateBridge.class ) ) {
```
Is there any strong need for the FQN?

gsmet replied:
Nope, just a copy/paste from the Lucene code, will fix.

gunnarmorling commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchCalendarBridge.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * Bridge a {@code java.util.Calendar} to a {@code String} using the ISO 8601 standard which is the default date format
+ * of Elasticsearch.
+ * UTC is used as time zone.
```
Did you consider to store the TZ offset instead (it's my understanding that ES supports formats with a TZ offset)? That way the calendar's TZ info can e.g. be shown when projecting the source document (or just a single field). That'd more be what I'd expect as a user.

gsmet replied:
Done.

gunnarmorling commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchCalendarBridge.java:
```diff
@@ -0,0 +1,70 @@
+		return super.objectToString( date );
+	}
+
+}
```
Nit-picking: By convention we do a line break at EOF. Actually we used to enforce it via CS but gave up on it because that check always was brittle.

gsmet replied:
Done.

gunnarmorling commented on engine/src/main/java/org/hibernate/search/bridge/IgnoreAnalyzerBridge.java:
```diff
@@ -0,0 +1,17 @@
+ *
+ * @author Guillaume Smet
+ */
+public interface IgnoreAnalyzerBridge {
```
Should this be a public contract right away? If so, should it be marked as experimental? I'd expect this to be morphed into a method on `FieldBridge` a la `boolean applyAnalyzers()` down the road instead of it being a marker interface?

gsmet replied:
Marked as experimental, moved to spi. Is it what you had in mind?

gunnarmorling commented:
@gsmet, looks good overall, see inline for some comments. Most notabily, I'd store the TZ offset in ES document fields for `Calendar` properties to facilitate projections (in the index, it will still be converted into millis-since-epoch).

gsmet commented:
@gunnarmorling I think I addressed all your comments. We should be ready to go!

gunnarmorling commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/util/impl/ElasticsearchDateHelper.java:
```diff
@@ -0,0 +1,92 @@
+	 * Limit a calendar resolution.
+	 *
+	 * @param resolution The desired resolution of the date to be returned
+	 * @return the calendar with all values more precise than <code>resolution</code> set to 0 or 1
```
A while ago we agreed on using `{@code ...}` instead of `<code>...`. I'll change it myself before applying it, just bringing it up to make sure we all are on the same page.

gunnarmorling commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/util/impl/ElasticsearchDateHelper.java:
```diff
@@ -0,0 +1,92 @@
+		return DatatypeConverter.printDateTime( c );
+	}
+
+	public static Calendar dateToCalendar(Date date) {
```
This one seems unused. I'll remove it.

gunnarmorling commented on elasticsearch/src/test/java/org/hibernate/search/backend/elasticsearch/test/ElasticsearchIndexMappingIT.java:
```diff
@@ -126,6 +131,7 @@ public void testMapping() throws Exception {
 				"{" +
 					"\"active\": true," +
 					"\"dateOfBirth\": \"1958-04-07T00:00:00Z\"," +
+					"\"subscriptionEndDate\": \"2016-06-07T00:00:00+02:00\"," +
```
Nice!

gunnarmorling commented:
Rebased, applied and added two more commits for some polishing. Let me know if you see any issues with those, @gsmet. Thanks!
